File: workspace/platf-bizz-prjs/drr-tnt-swfabrik-europe/food-market-03-service/internal/application/service.go
package application

import (
	"context"
	"errors"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/darueira/foodmarket/service03/internal/domain"
	"github.com/darueira/foodmarket/service03/internal/port"
	"github.com/google/uuid"
)

type FoodTradingService struct {
	persistence    port.FoodTradingPersistencePort
	publisher      port.FoodTradingEventPublisherPort
	sseBroadcaster port.FoodTradingSseBroadcasterPort
	marketID       string
}

func NewFoodTradingService(
	p port.FoodTradingPersistencePort,
	pub port.FoodTradingEventPublisherPort,
	sse port.FoodTradingSseBroadcasterPort,
	marketID string,
) *FoodTradingService {
	return &FoodTradingService{
		persistence:    p,
		publisher:      pub,
		sseBroadcaster: sse,
		marketID:       marketID,
	}
}

func (s *FoodTradingService) CreateTrading(ctx context.Context, cmd domain.CreateFoodTradingCommand) (*domain.FoodTrading, error) {
	log.Printf("[Go 03] Creating food trading: %s by %s", cmd.ItemName, cmd.TraderName)

	trading := &domain.FoodTrading{
		MarketID:   s.marketID,
		ItemName:   cmd.ItemName,
		Quantity:   cmd.Quantity,
		UnitPrice:  cmd.UnitPrice,
		TraderName: cmd.TraderName,
		Status:     "CONFIRMED",
		CreatedAt:  time.Now().UTC(),
	}

	if err := trading.ValidateAndCalculate(); err != nil {
		return nil, err
	}

	// 1. Persist to PostgreSQL (F02)
	saved, err := s.persistence.Save(ctx, trading)
	if err != nil {
		return nil, fmt.Errorf("failed to persist trading: %w", err)
	}

	// 2. Publish to RabbitMQ Topic (F03)
	event := domain.FoodTradingEvent{
		EventID:    uuid.New().String(),
		EventType:  "FOOD_TRADING_CREATED",
		TradingID:  saved.TradingID,
		MarketID:   saved.MarketID,
		ItemName:   saved.ItemName,
		Quantity:   saved.Quantity,
		UnitPrice:  saved.UnitPrice,
		TotalPrice: saved.TotalPrice,
		TraderName: saved.TraderName,
		Status:     saved.Status,
		Timestamp:  time.Now().UTC(),
	}

	if err := s.publisher.PublishEvent(ctx, event); err != nil {
		log.Printf("[Go 03] Warning: failed to publish to RabbitMQ: %v", err)
	}

	// 3. Broadcast to SSE (F01.3 / F04.2.3)
	s.sseBroadcaster.Broadcast(*saved)

	return saved, nil
}

func (s *FoodTradingService) ListTradings(ctx context.Context) ([]domain.FoodTrading, error) {
	return s.persistence.FindAll(ctx)
}

func (s *FoodTradingService) SubscribeStream() <-chan domain.FoodTrading {
	return s.sseBroadcaster.Subscribe()
}

func (s *FoodTradingService) UnsubscribeStream(ch <-chan domain.FoodTrading) {
	s.sseBroadcaster.Unsubscribe(ch)
}

func (s *FoodTradingService) ProcessIncomingEvent(ctx context.Context, event domain.FoodTradingEvent) error {
	log.Printf("[Go 03] Consuming RabbitMQ message event: %s", event.TradingID)

	if event.Quantity <= 0 || event.UnitPrice <= 0 {
		log.Printf("[Go 03] Discarded invalid event: %s", event.TradingID)
		return errors.New("invalid event payload")
	}

	// Check idempotency
	existing, err := s.persistence.FindByTradingID(ctx, event.TradingID)
	if err == nil && existing != nil {
		log.Printf("[Go 03] Trading %s already exists in DB, skipping duplicate persist", event.TradingID)
		return nil
	}

	mktID := event.MarketID
	if strings.TrimSpace(mktID) == "" {
		mktID = s.marketID
	}

	trading := &domain.FoodTrading{
		TradingID:  event.TradingID,
		MarketID:   mktID,
		ItemName:   event.ItemName,
		Quantity:   event.Quantity,
		UnitPrice:  event.UnitPrice,
		TraderName: event.TraderName,
		Status:     "CONFIRMED",
		CreatedAt:  event.Timestamp,
	}

	if err := trading.ValidateAndCalculate(); err != nil {
		return err
	}

	saved, err := s.persistence.Save(ctx, trading)
	if err != nil {
		return err
	}

	s.sseBroadcaster.Broadcast(*saved)
	return nil
}

File: apps/examples/tenant-xpto/food-market-04-service/app/application/service.py
import logging
from decimal import Decimal
from typing import List, AsyncGenerator
from app.domain.models import FoodTrading, FoodTradingEvent, CreateFoodTradingCommand
from app.port.inbound import FoodTradingUseCase
from app.port.outbound import FoodTradingPersistencePort, FoodTradingEventPublisherPort, FoodTradingSseBroadcasterPort

logger = logging.getLogger(__name__)

class FoodTradingService(FoodTradingUseCase):
    def __init__(
        self,
        persistence: FoodTradingPersistencePort,
        publisher: FoodTradingEventPublisherPort,
        sse_broadcaster: FoodTradingSseBroadcasterPort,
        market_id: str,
    ):
        self.persistence = persistence
        self.publisher = publisher
        self.sse_broadcaster = sse_broadcaster
        self.market_id = market_id

    async def create_trading(self, cmd: CreateFoodTradingCommand) -> FoodTrading:
        logger.info(f"[Python 04] Creating food trading: {cmd.item_name} by {cmd.trader_name}")
        trading = FoodTrading(
            market_id=self.market_id,
            item_name=cmd.item_name,
            quantity=cmd.quantity,
            unit_price=cmd.unit_price,
            trader_name=cmd.trader_name,
        )
        trading.validate_and_calculate()

        # 1. Persist (F02)
        saved = await self.persistence.save(trading)

        # 2. Publish to RabbitMQ Topic (F03)
        event = FoodTradingEvent(
            trading_id=saved.trading_id,
            market_id=saved.market_id,
            item_name=saved.item_name,
            quantity=saved.quantity,
            unit_price=saved.unit_price,
            total_price=saved.total_price,
            trader_name=saved.trader_name,
            status=saved.status,
            timestamp=saved.created_at,
        )
        await self.publisher.publish_event(event)

        # 3. Broadcast to SSE (F01.3 / F04.2.3)
        await self.sse_broadcaster.broadcast(saved)

        return saved

    async def list_tradings(self) -> List[FoodTrading]:
        return await self.persistence.find_all()

    async def subscribe_stream(self) -> AsyncGenerator[FoodTrading, None]:
        # Generator handled by SSE Broadcaster queue
        pass

    async def process_incoming_event(self, event: FoodTradingEvent) -> None:
        logger.info(f"[Python 04] Consuming RabbitMQ message event: {event.trading_id}")

        if event.quantity <= Decimal("0.00") or event.unit_price <= Decimal("0.00"):
            logger.warning(f"Discarding invalid event: {event.trading_id}")
            return

        existing = await self.persistence.find_by_trading_id(event.trading_id)
        if existing is not None:
            logger.info(f"Trading {event.trading_id} already exists in DB, skipping duplicate persist")
            return

        trading = FoodTrading(
            trading_id=event.trading_id,
            market_id=event.market_id or self.market_id,
            item_name=event.item_name,
            quantity=event.quantity,
            unit_price=event.unit_price,
            trader_name=event.trader_name,
            status="CONFIRMED",
            created_at=event.timestamp,
        )
        trading.validate_and_calculate()

        saved = await self.persistence.save(trading)
        await self.sse_broadcaster.broadcast(saved)
